File: dodam-application/dodam-rest-api/src/main/java/b1nd/dodam/restapi/wakeupsong/application/WakeupSongUseCase.java
package b1nd.dodam.restapi.wakeupsong.application;

import b1nd.dodam.core.exception.ExceptionCode;
import b1nd.dodam.core.exception.global.GlobalExceptionCode;
import b1nd.dodam.domain.rds.member.entity.Member;
import b1nd.dodam.domain.rds.wakeupsong.entity.WakeupSong;
import b1nd.dodam.domain.rds.wakeupsong.exception.UnsupportedVideoTypeException;
import b1nd.dodam.domain.rds.wakeupsong.exception.WakeupSongAlreadyCreatedException;
import b1nd.dodam.domain.rds.wakeupsong.exception.WakeupSongExceptionCode;
import b1nd.dodam.domain.rds.wakeupsong.exception.WakeupSongUrlMalformedException;
import b1nd.dodam.domain.rds.wakeupsong.service.WakeupSongService;
import b1nd.dodam.melon.chart.client.MelonChartClient;
import b1nd.dodam.melon.chart.client.data.res.ChartRes;
import b1nd.dodam.restapi.auth.infrastructure.security.support.MemberAuthenticationHolder;
import b1nd.dodam.restapi.support.data.ErrorResponseEntity;
import b1nd.dodam.restapi.support.data.Response;
import b1nd.dodam.restapi.support.data.ResponseData;
import b1nd.dodam.restapi.wakeupsong.application.data.req.ApplyWakeupSongBySearchReq;
import b1nd.dodam.restapi.wakeupsong.application.data.res.WakeupSongRes;
import b1nd.dodam.restapi.wakeupsong.application.support.HtmlConverter;
import b1nd.dodam.restapi.wakeupsong.application.support.YoutubeApiUtil;
import b1nd.dodam.youtube.video.client.YoutubeVideoClient;
import b1nd.dodam.youtube.video.client.data.YoutubeApiRes;
import b1nd.dodam.youtube.video.client.data.YoutubeRes;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

@Component
@RequiredArgsConstructor
public class WakeupSongUseCase {

    private final WakeupSongService wakeupSongService;
    private final YoutubeVideoClient youtubeVideoClient;
    private final MelonChartClient melonChartClient;
    private final MemberAuthenticationHolder memberAuthenticationHolder;

    @Transactional(readOnly = true)
    public ResponseData<List<WakeupSongRes>> getAllowedWakeupSong(int year, int month, int day){
        List<WakeupSong> wakeupSongList = wakeupSongService.getByPlayDate(year, month, day);
        return ResponseData.ok("승인된 기상송 조회 성공", WakeupSongRes.of(wakeupSongList));
    }

    @Transactional(readOnly = true)
    public ResponseData<List<WakeupSongRes>> getPendingWakeupSong(){
        List<WakeupSong> wakeupSongList = wakeupSongService.getPendingWakeupSong();
        return ResponseData.ok("승인 대기 중인 기상송 조회 성공", WakeupSongRes.of(wakeupSongList));
    }

    @Transactional(readOnly = true)
    public ResponseData<List<WakeupSongRes>> getMyWakeupSong(){
        Member member = memberAuthenticationHolder.current();
        List<WakeupSong> wakeupSongList = wakeupSongService.getMyWakeupSong(member);
        return ResponseData.ok("자신이 신청한 기상송 조회 성공", WakeupSongRes.of(wakeupSongList));
    }

    @Transactional(rollbackFor = Exception.class)
    public CompletableFuture<Response> createWakeupSong(String videoUrl) {
        Member member = verifyAlreadyAppliedFromSession();
        String videoId = YoutubeApiUtil.getVideoId(videoUrl);

        return youtubeVideoClient.getVideo(videoId)
                .thenApply(videoResponse -> {
                    YoutubeApiRes.Snippet snippet = videoResponse.getItems().get(0).getSnippet();
                    checkValidVideoType(snippet.getTitle());
                    buildAndSaveWakeupSong(snippet, videoId, videoUrl, member);
                    return Response.created("기상송 신청 성공");
                }).exceptionally(this::handleExceptionOnCreateWakeupSong);
    }


    @Transactional(rollbackFor = Exception.class)
    public CompletableFuture<Response> createWakeupSongByYoutubeSearch(ApplyWakeupSongBySearchReq req){
        Member member = verifyAlreadyAppliedFromSession();
        return CompletableFuture.supplyAsync(() -> {
            YoutubeApiRes.SearchItem searchItem = youtubeVideoClient.searchVideoByKeyword(
                    req.title() + " " + req.artist(), 1).getItems().get(0);
            checkValidVideoType(searchItem.getSnippet().getTitle());
            buildAndSaveWakeupSong(searchItem.getSnippet(), searchItem.getId().getVideoId(),
                    "https://www.youtube.com/watch?v=" + searchItem.getId().getVideoId(), member);
            return Response.created("유튜브 검색을 통한 기상송 신청 성공");
        }).exceptionally(this::handleExceptionOnCreateWakeupSong);
    }

    private Response handleExceptionOnCreateWakeupSong(Throwable e) {
        Throwable cause = (e instanceof CompletionException) ? e.getCause() : e;
        ExceptionCode exceptionCode = mapExceptionToCodeOnCreateWakeupSong(cause);
        return ErrorResponseEntity.of(exceptionCode.getStatus(), exceptionCode.getExceptionName(), exceptionCode.getMessage());
    }

    private ExceptionCode mapExceptionToCodeOnCreateWakeupSong(Throwable cause) {
        if (cause instanceof WakeupSongUrlMalformedException) {
            return WakeupSongExceptionCode.URL_MALFORMED;
        } else if (cause instanceof UnsupportedVideoTypeException) {
            return WakeupSongExceptionCode.UNSUPPORTED_TYPE;
        } else if(cause instanceof WakeupSongAlreadyCreatedException){
            return WakeupSongExceptionCode.ALREADY_APPLIED;
        }else {
            return GlobalExceptionCode.INTERNAL_SERVER;
        }
    }

    private void buildAndSaveWakeupSong(YoutubeApiRes.Snippet snippet, String videoId, String videoUrl, Member member){
         WakeupSong wakeupSong = WakeupSong.builder()
                .videoId(videoId)
                .videoTitle(HtmlConverter.of(snippet.getTitle()))
                .videoUrl(videoUrl)
                .channelTitle(snippet.getChannelTitle())
                .thumbnailUrl(YoutubeApiUtil.getThumbnailUrl(snippet).getUrl())
                .member(member)
                .build();
        wakeupSongService.saveWakeupSong(wakeupSong);
    }

    private void checkValidVideoType(String title){
        if(title.contains("MV")){
            throw new UnsupportedVideoTypeException();
        }
    }

    private Member verifyAlreadyAppliedFromSession(){
        Member member = memberAuthenticationHolder.current();
        if (wakeupSongService.existsByMemberAndCreatedAt(member)){
            throw new WakeupSongAlreadyCreatedException();
        }
        return member;
    }

    @Transactional(rollbackFor = Exception.class)
    public Response allow(int id){
        WakeupSong wakeupSong = wakeupSongService.getById(id);
        wakeupSong.allow(memberAuthenticationHolder.current());
        return Response.ok("기상송 승인 성공");
    }

    @Transactional(rollbackFor = Exception.class)
    public Response deny(int id){
        WakeupSong wakeupSong = wakeupSongService.getById(id);
        wakeupSong.deny();
        return Response.ok("기상송 거절 성공");
    }

    @Transactional(rollbackFor = Exception.class)
    public Response delete(int id){
        WakeupSong wakeupSong = wakeupSongService.getById(id);
        wakeupSongService.delete(wakeupSong);
        return Response.ok("기상송 삭제 성공");
    }

    @Transactional(rollbackFor = Exception.class)
    public Response deleteMyWakeupSong(int id) {
        WakeupSong wakeupSong = wakeupSongService.getById(id);
        wakeupSong.isApplicant(memberAuthenticationHolder.current());
        wakeupSongService.delete(wakeupSong);
        return Response.ok("기상송 신청 취소 성공");
    }

    @Transactional(readOnly = true)
    public CompletableFuture<ResponseData<List<ChartRes>>> getChartList() {
        return melonChartClient.getList().thenApply(res ->  ResponseData.ok("차트 조회 성공", res));
    }

    @Transactional(readOnly = true)
    public ResponseData<List<YoutubeRes>> getYoutubeList(String keyword) {
        List<YoutubeRes> videoList = youtubeVideoClient.searchVideoByKeyword(keyword, 5).getItems().stream()
                .map(YoutubeApiUtil::getYoutubeRes).toList();
        return ResponseData.ok("유튜브 검색을 통한 기상송 조회 성공", videoList);
    }

}

File: services/service-wakeup-song/src/main/kotlin/com/b1nd/dodamdodam/wakeupsong/application/WakeupSongUseCase.kt
package com.b1nd.dodamdodam.wakeupsong.application

import com.b1nd.dodamdodam.core.common.data.FileDownloadResponse
import com.b1nd.dodamdodam.core.common.data.Response
import com.b1nd.dodamdodam.core.security.passport.holder.PassportHolder
import com.b1nd.dodamdodam.core.security.passport.requireUserId
import com.b1nd.dodamdodam.wakeupsong.application.data.request.ApplyWakeupSongByKeywordRequest
import com.b1nd.dodamdodam.wakeupsong.application.data.request.ApplyWakeupSongRequest
import com.b1nd.dodamdodam.wakeupsong.application.data.response.MelonChartResponse
import com.b1nd.dodamdodam.wakeupsong.application.data.response.WakeupSongResponse
import com.b1nd.dodamdodam.wakeupsong.application.data.response.YouTubeSearchResponse
import com.b1nd.dodamdodam.wakeupsong.application.data.toEntity
import com.b1nd.dodamdodam.wakeupsong.application.data.toResponse
import com.b1nd.dodamdodam.wakeupsong.domain.wakeupsong.exception.WakeupSongNotOwnerException
import com.b1nd.dodamdodam.wakeupsong.domain.wakeupsong.exception.YouTubeVideoNotFoundException
import com.b1nd.dodamdodam.wakeupsong.domain.wakeupsong.service.WakeupSongService
import com.b1nd.dodamdodam.wakeupsong.infrastructure.melon.MelonChartClient
import com.b1nd.dodamdodam.wakeupsong.infrastructure.youtube.YtDlpClient
import com.b1nd.dodamdodam.wakeupsong.infrastructure.youtube.YouTubeClient
import org.springframework.core.io.Resource
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.nio.file.Files

@Component
@Transactional(rollbackFor = [Exception::class])
class WakeupSongUseCase(
    private val wakeupSongService: WakeupSongService,
    private val youTubeClient: YouTubeClient,
    private val melonChartClient: MelonChartClient,
    private val ytDlpClient: YtDlpClient
) {

    fun applyWakeupSong(request: ApplyWakeupSongRequest): Response<Any> {
        val userId = PassportHolder.current().requireUserId()
        val videoId = YouTubeClient.extractVideoId(request.videoUrl)
            ?: throw YouTubeVideoNotFoundException()
        val videoInfo = youTubeClient.getVideoInfo(videoId)
        val entity = videoInfo.toEntity(userId)
        wakeupSongService.save(entity)
        return Response.created("기상송이 신청되었어요.")
    }

    fun applyByKeyword(request: ApplyWakeupSongByKeywordRequest): Response<Any> {
        val userId = PassportHolder.current().requireUserId()
        val keyword = "${request.artist} ${request.title}"
        val results = youTubeClient.search(keyword)
        val videoInfo = results.firstOrNull() ?: throw YouTubeVideoNotFoundException()
        val entity = videoInfo.toEntity(userId)
        wakeupSongService.save(entity)
        return Response.created("기상송이 신청되었어요.")
    }

    @Transactional(readOnly = true)
    fun getMyWakeupSongs(): Response<List<WakeupSongResponse>> {
        val userId = PassportHolder.current().requireUserId()
        val songs = wakeupSongService.getByUserId(userId).map { it.toResponse() }
        return Response.ok("내 기상송을 조회했어요.", songs)
    }

    @Transactional(readOnly = true)
    fun getAllowed(year: Int, month: Int, day: Int): Response<List<WakeupSongResponse>> {
        val songs = wakeupSongService.getAllowed(year, month, day).map { it.toResponse() }
        return Response.ok("승인된 기상송을 조회했어요.", songs)
    }

    @Transactional(readOnly = true)
    fun getPending(): Response<List<WakeupSongResponse>> {
        val songs = wakeupSongService.getPending().map { it.toResponse() }
        return Response.ok("대기 중인 기상송을 조회했어요.", songs)
    }

    fun allow(id: Long): Response<Any> {
        val wakeupSong = wakeupSongService.getById(id)
        wakeupSong.allow()
        wakeupSongService.save(wakeupSong)
        return Response.ok("기상송이 승인되었어요.")
    }

    fun deny(id: Long): Response<Any> {
        val wakeupSong = wakeupSongService.getById(id)
        wakeupSong.deny()
        wakeupSongService.save(wakeupSong)
        return Response.ok("기상송이 거절되었어요.")
    }

    fun deleteMyWakeupSong(id: Long): Response<Any> {
        val userId = PassportHolder.current().requireUserId()
        val wakeupSong = wakeupSongService.getById(id)
        if (wakeupSong.userId != userId) {
            throw WakeupSongNotOwnerException()
        }
        wakeupSongService.delete(wakeupSong)
        return Response.ok("기상송이 삭제되었어요.")
    }

    @Transactional(readOnly = true)
    fun searchYouTube(keyword: String): Response<List<YouTubeSearchResponse>> {
        val results = youTubeClient.search(keyword)
        return Response.ok("유튜브 검색 결과를 조회했어요.", results)
    }

    @Transactional(readOnly = true)
    fun getMelonChart(): Response<List<MelonChartResponse>> {
        val chart = melonChartClient.getChart()
        return Response.ok("멜론 차트를 조회했어요.", chart)
    }

    fun downloadMp3(url: String): ResponseEntity<Resource> {
        val videoId = YouTubeClient.extractVideoId(url)
            ?: throw YouTubeVideoNotFoundException()
        val videoInfo = youTubeClient.getVideoInfo(videoId)
        val file = ytDlpClient.downloadMp3(url)
        val bytes = Files.readAllBytes(file)
        Files.deleteIfExists(file)
        val response = FileDownloadResponse(bytes, "${videoInfo.videoTitle}.mp3", MediaType.parseMediaType("audio/mpeg"))
        return Response.toFileResponseEntity(response)
    }
}
